lib/Parse/ParseType.cpp

```diff
@@ -646,25 +675,25 @@ Parser::parseTypeIdentifier(bool isParsingQualifiedDeclName) {
     // If parsing qualified decl name and encountered a token like `.+` in
     // `Float.+`, consume the period so that `+` is treated correctly as a
     // standalone token.
-    } else if (isParsingQualifiedDeclName && startsWithSymbol(Tok, '.')) {
-      if (Tok.getLength() != 1 || !peekToken().is(tok::identifier)) {
-        consumeStartingCharacterOfCurrentToken(tok::period);
-        continue;
-      }
+    } else if (isParsingQualifiedDeclName && startsWithSymbol(Tok, '.') &&
+               Tok.getLength() != 1) {
+      consumeStartingCharacterOfCurrentToken(tok::period);
+      continue;
     }
     break;
   }
 
   if (isParsingQualifiedDeclName) {
-    // If parsing qualified decl name and the current token is not an operator,
-    // backtrack before parsing the last identifier.
-    if (Tok.isNotAnyOperator()) {
-      restoreParserPosition(backtrackingPosition);
-      if (!ComponentsR.empty())
-        ComponentsR.pop_back();
-    }
-    if (ComponentsR.empty())
+    // If components are empty, discard the syntax parsing context.
+    if (ComponentsR.empty()) {
+      IdentTypeCtxt.setDiscard();
```

Comment:
Any reason we're using `setDiscard()` here instead of using `setTransparent()`? We usually only use `setDiscard()` in back tracking context. One goal of syntax tress is we can round-trip print even invalid code. Otherwise, LGTM!

Reply:
Good point, thanks! I didn't expect `setTransparent` to work in place of `setDiscard`, but it did which is nice. Done in https://github.com/apple/swift/pull/17725/commits/5a64f20b9524636216f575163f60e92537c170fc.

In fact, backtracking is being performed here (I added the `isParsingQualifiedDeclName` argument to `parseTypeIdentifier` to enable backtracking the last type component, [explanation here](https://github.com/apple/swift/pull/17324#issuecomment-398514702)).